WIP: Add tests for HttpPostContext content-type

mlevesquedion opened the pull request:
I've added a couple tests which I think cover most of the functionality. I couldn't get the first way (header) to work.

I don't believe this is ready to merge, hence the "WIP" in the title. Mostly looking to see whether I'm going in the right direction or not.

Cheers!

rybalkinsd commented:
hey @mlevesquedion !

You're on the right way. However, it's also important to check the case:

```kotlin
httpPost {
    host = 
    path =
    header {
        "Content-Type" to ...
    }
    body(type) {}
}
```
OkHttp has an [issue](https://github.com/square/okhttp/issues/2099) related to the topic. However header "Content-Type" might be ignored in okhttp (important to check). 

Also, it's important to add a note to `body { }` DSL about content type resolution in case of multi-declaration.

mlevesquedion commented:
Hey there!

Thanks for your feedback! I'll add the missing tests and the note later today.

mlevesquedion commented:
Here's an update. If I understood you correctly, this should do it.

Summary of my changes :
1. Get content type from header as default argument for body
2. Add tests covering remaining features, removed one test which was not necessary
3. Modified the README to show how to set the content type in the header and inform users about content type resolution

rybalkinsd commented:
@mlevesquedion 
I might be not concrete enough with the details.
As mentioned in #62 task scope, I expected only tests and readme.md to be changed.
According to this I suggest to leave only related changes in this PR. 
/src changes better to provide in a separate PR. ( it would be much easier for release. Test and docs could be merged instantly, but sources will be merged in next release brunch)

Btw, did you check the behavior of okhttp3 for case when post request has both body with Content-Type and header with different type?

rybalkinsd commented on src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HttpContext.kt:
```diff
@@ -66,12 +66,16 @@ sealed class HttpContext(private val method: Method = GET) : IHttpContext {
     private var body: RequestBody = RequestBody.create(null, byteArrayOf())
 
-    fun body(contentType: String? = null, init: BodyContext.() -> RequestBody) {
+    fun body(contentType: String? = getHeaderContentType(), init: BodyContext.() -> RequestBody) {
```
What if 
Body declared before header?

`httpPost { body { } header { } }`

rybalkinsd commented on src/test/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HttpPostContextTest.kt:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Test
+    fun `should override custom content type with type inside body`() {
```
This test proves that changes in readme are not fully correct

rybalkinsd commented on README.md:
````diff
@@ -162,6 +162,25 @@ httpPost {
+You may also specify a custom content type in the request's header :
+```kotlin
+httpPost {
+    header { "Content-Type" to "text/html" }
````
Not sure that we need to recommend this option.
It’s better to mention in body javadoc.

rybalkinsd commented on src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HeaderContext.kt:
```diff
@@ -18,4 +18,8 @@ class HeaderContext {
 
     internal fun forEach(action: (k: String, v: Any) -> Unit) = map.forEach(action)
+
+    fun getContentType(): String? {
```
As I previously commented, suggest to move /src changes to separate PR

rybalkinsd replied:
Also looks like this function better to be private

rybalkinsd commented on README.md:
````diff
@@ -162,6 +162,25 @@ httpPost {
+
+Content type is set according to the following priority levels (higher is prioritized) :
+
+1. Header : ```kotlin ... header { "Content-Type" to "$myContenType" } ... ```  
````
Please double check the priority ordering

rybalkinsd replied:
In case we move /src changes to separate PR, will header be first?

mlevesquedion commented:
Hi @rybalkinsd,

Thank you for the comments! I haven't been able to look at this in the past few days, but I might have some time to look at it this weekend. In the meantime, if you want to go with a different approach, that is totally okay with me. Cheers!

rybalkinsd commented:
@mlevesquedion Take your time! We're not urgent with this task. In my mind we can iteratively get to the production solution soon!

doyaaaaaken commented:
This PR is stale because issue #62 is already resolved.
So, maybe we should close this PR.
